pretyping/coercion.ml

```diff
@@ -367,36 +353,98 @@ let saturate_evd env evd =
   Typeclasses.resolve_typeclasses
     ~filter:Typeclasses.no_goals ~split:true ~fail:false env evd
 
+type coercion_trace =
+  | IdCoe
+  | PrimProjCoe of {
+      proj : Projection.Repr.t;
+      head : econstr;
+      args : econstr list;
+      previous : coercion_trace;
+    }
+  | Coe of {
+      head : econstr;
+      args : econstr list;
+      previous : coercion_trace;
+    }
+  | ProdCoe of { na : Name.t binder_annot; ty : econstr; dom : coercion_trace; body : coercion_trace }
+
+let empty_coercion_trace = IdCoe
+
+(* similar to iterated apply_coercion_args *)
+let rec reapply_coercions sigma trace c = match trace with
+  | IdCoe -> c
+  | PrimProjCoe { proj; head; args; previous } ->
+    let c = reapply_coercions sigma previous c in
+    let args = args@[c] in
+    applist (mkProj (Projection.make proj false, head), args)
+  | Coe {head; args; previous} ->
+    let c = reapply_coercions sigma previous c in
+    let args = args@[c] in
+    applist (head, args)
+  | ProdCoe { na; ty; dom; body } ->
+    let x = reapply_coercions sigma dom (mkRel 1) in
+    let c = beta_applist sigma (lift 1 c, [x]) in
+    let c = reapply_coercions sigma body c in
+    mkLambda (na, ty, c)
+
 (* Apply coercion path from p to hj; raise NoCoercion if not applicable *)
 let apply_coercion env sigma p hj typ_cl =
-  let j,t,evd =
+  let j,t,trace,evd =
     List.fold_left
-      (fun (ja,typ_cl,sigma) i ->
+      (fun (ja,typ_cl,trace,sigma) i ->
          let isid = i.coe_is_identity in
          let isproj = i.coe_is_projection in
          let sigma, c = new_global sigma i.coe_value in
          let typ = Retyping.get_type_of env sigma c in
          let fv = make_judge c typ in
-         let argl = (class_args_of env sigma typ_cl)@[ja.uj_val] in
-         let sigma, jres =
-           apply_coercion_args env sigma true isproj argl fv
+         let argl = class_args_of env sigma typ_cl in
+         let trace =
+           if isid then trace
+           else match isproj with
+           | None -> Coe {head=fv.uj_val;args=argl;previous=trace}
+           | Some proj ->
+             let args = List.skipn (Projection.Repr.npars proj) argl in
+             let head, args = match args with [] -> assert false | hd :: tl -> hd, tl in
```

Comment:
The skipn (ie the ine this is copied from, in apply_coercion_args) has been broken for a while (https://github.com/coq/coq/issues/11366)
However this introduces a new break when argl is empty (in apply_coercion_args it is never empty since we pass `argl@[ja.uj_val]`), this is what @gares is talking about in his comment on the reapply_coercions code.
IMO we should not split head/args in the PrimProjCoe data, do inside reapply_coercions.
Not sure what the fix for the skipn bug will be, IMO we should also do it in reapply_coercions since that may be easier to make an uniform fix for though.